File: client/Assets/Protocol/Wrap/Response.cs
using System.Collections.Generic;
using System.Linq;

namespace Protocol.Response
{
    public enum Identity
    {
        CREATE_ROOM,
        JOIN_ROOM,
        LEAVE_ROOM,
        KICK_ROOM,
        KICKED_ROOM
    }

    public class CreateRoom : IProtocol
    {
        public uint Identity => (uint)Protocol.Response.Identity.CREATE_ROOM;

        public uint Id { get; set; }

        public CreateRoom()
        { }

        public CreateRoom(FlatBuffer.Response.CreateRoom obj)
        {
            this.Id = obj.Id;
        }

        public FlatBuffers.Offset<FlatBuffer.Response.CreateRoom> ToFlatBuffer(FlatBuffers.FlatBufferBuilder builder)
        {
            var _id = this.Id;

            return FlatBuffer.Response.CreateRoom.CreateCreateRoom(builder, _id);
        }

        public byte[] Serialize()
        {
            var builder = new FlatBuffers.FlatBufferBuilder(512);
            builder.Finish(this.ToFlatBuffer(builder).Value);
            return builder.DataBuffer.ToSizedArray();
        }

        public static CreateRoom Deserialize(byte[] bytes)
        {
            return new CreateRoom(FlatBuffer.Response.CreateRoom.GetRootAsCreateRoom(new FlatBuffers.ByteBuffer(bytes)));
        }
    }

    public class JoinRoom : IProtocol
    {
        public uint Identity => (uint)Protocol.Response.Identity.JOIN_ROOM;

        public List<ulong> Users { get; set; }

        public JoinRoom()
        { }

        public JoinRoom(FlatBuffer.Response.JoinRoom obj)
        {
            this.Users = Enumerable.Range(0, obj.UsersLength).Select(x => (ulong)System.Convert.ChangeType(x, typeof(ulong))).ToList();
        }

        public FlatBuffers.Offset<FlatBuffer.Response.JoinRoom> ToFlatBuffer(FlatBuffers.FlatBufferBuilder builder)
        {
            var _users = FlatBuffer.Response.JoinRoom.CreateUsersVector(builder, this.Users.ToArray());

            return FlatBuffer.Response.JoinRoom.CreateJoinRoom(builder, _users);
        }

        public byte[] Serialize()
        {
            var builder = new FlatBuffers.FlatBufferBuilder(512);
            builder.Finish(this.ToFlatBuffer(builder).Value);
            return builder.DataBuffer.ToSizedArray();
        }

        public static JoinRoom Deserialize(byte[] bytes)
        {
            return new JoinRoom(FlatBuffer.Response.JoinRoom.GetRootAsJoinRoom(new FlatBuffers.ByteBuffer(bytes)));
        }
    }

    public class LeaveRoom : IProtocol
    {
        public uint Identity => (uint)Protocol.Response.Identity.LEAVE_ROOM;

        

        public LeaveRoom()
        { }

        public LeaveRoom(FlatBuffer.Response.LeaveRoom obj)
        {
            
        }

        public FlatBuffers.Offset<FlatBuffer.Response.LeaveRoom> ToFlatBuffer(FlatBuffers.FlatBufferBuilder builder)
        {
            FlatBuffer.Response.LeaveRoom.StartLeaveRoom(builder);
            return FlatBuffer.Response.LeaveRoom.EndLeaveRoom(builder);
        }

        public byte[] Serialize()
        {
            var builder = new FlatBuffers.FlatBufferBuilder(512);
            builder.Finish(this.ToFlatBuffer(builder).Value);
            return builder.DataBuffer.ToSizedArray();
        }

        public static LeaveRoom Deserialize(byte[] bytes)
        {
            return new LeaveRoom(FlatBuffer.Response.LeaveRoom.GetRootAsLeaveRoom(new FlatBuffers.ByteBuffer(bytes)));
        }
    }

    public class KickRoom : IProtocol
    {
        public uint Identity => (uint)Protocol.Response.Identity.KICK_ROOM;

        public bool Success { get; set; }

        public KickRoom()
        { }

        public KickRoom(FlatBuffer.Response.KickRoom obj)
        {
            this.Success = obj.Success;
        }

        public FlatBuffers.Offset<FlatBuffer.Response.KickRoom> ToFlatBuffer(FlatBuffers.FlatBufferBuilder builder)
        {
            var _success = this.Success;

            return FlatBuffer.Response.KickRoom.CreateKickRoom(builder, _success);
        }

        public byte[] Serialize()
        {
            var builder = new FlatBuffers.FlatBufferBuilder(512);
            builder.Finish(this.ToFlatBuffer(builder).Value);
            return builder.DataBuffer.ToSizedArray();
        }

        public static KickRoom Deserialize(byte[] bytes)
        {
            return new KickRoom(FlatBuffer.Response.KickRoom.GetRootAsKickRoom(new FlatBuffers.ByteBuffer(bytes)));
        }
    }

    public class KickedRoom : IProtocol
    {
        public uint Identity => (uint)Protocol.Response.Identity.KICKED_ROOM;

        

        public KickedRoom()
        { }

        public KickedRoom(FlatBuffer.Response.KickedRoom obj)
        {
            
        }

        public FlatBuffers.Offset<FlatBuffer.Response.KickedRoom> ToFlatBuffer(FlatBuffers.FlatBufferBuilder builder)
        {
            FlatBuffer.Response.KickedRoom.StartKickedRoom(builder);
            return FlatBuffer.Response.KickedRoom.EndKickedRoom(builder);
        }

        public byte[] Serialize()
        {
            var builder = new FlatBuffers.FlatBufferBuilder(512);
            builder.Finish(this.ToFlatBuffer(builder).Value);
            return builder.DataBuffer.ToSizedArray();
        }

        public static KickedRoom Deserialize(byte[] bytes)
        {
            return new KickedRoom(FlatBuffer.Response.KickedRoom.GetRootAsKickedRoom(new FlatBuffers.ByteBuffer(bytes)));
        }
    }
}

File: server/src/protocol/Response/Response.go
package response

import (
	"encoding/binary"
	"protocol"
	source "protocol/FlatBuffer/Response"

	flatbuffers "github.com/google/flatbuffers/go"
)

func Deserialize(size uint32, bytes []byte) protocol.Protocol {
	offset := 0
	identity := binary.LittleEndian.Uint32(bytes[:4])
	offset += 4

	payload := bytes[offset : offset+int(size)]
	switch identity {
	case CREATE_ROOM:
		x := &CreateRoom{}
		return x.Deserialize(payload)

	case JOIN_ROOM:
		x := &JoinRoom{}
		return x.Deserialize(payload)

	case LEAVE_ROOM:
		x := &LeaveRoom{}
		return x.Deserialize(payload)

	case KICK_ROOM:
		x := &KickRoom{}
		return x.Deserialize(payload)

	case KICKED_ROOM:
		x := &KickedRoom{}
		return x.Deserialize(payload)

	case DESTROYED_ROOM:
		x := &DestroyedRoom{}
		return x.Deserialize(payload)

	}

	return nil
}

func Text(p protocol.Protocol) string {
	switch p.(type) {
	case *CreateRoom:
		return "CREATE_ROOM"

	case *JoinRoom:
		return "JOIN_ROOM"

	case *LeaveRoom:
		return "LEAVE_ROOM"

	case *KickRoom:
		return "KICK_ROOM"

	case *KickedRoom:
		return "KICKED_ROOM"

	case *DestroyedRoom:
		return "DESTROYED_ROOM"
	}
	return ""
}

const (
	CREATE_ROOM = iota
	JOIN_ROOM
	LEAVE_ROOM
	KICK_ROOM
	KICKED_ROOM
	DESTROYED_ROOM
)

type CreateRoom struct {
	Id    string
	Error uint32
}

func (obj *CreateRoom) create(builder *flatbuffers.Builder) flatbuffers.UOffsetT {
	_id := builder.CreateString(obj.Id)

	source.CreateRoomStart(builder)
	source.CreateRoomAddId(builder, _id)
	source.CreateRoomAddError(builder, obj.Error)

	return source.CreateRoomEnd(builder)
}

func (obj *CreateRoom) parse(x *source.CreateRoom) *CreateRoom {
	obj.Id = string(x.Id())
	obj.Error = x.Error()

	return obj
}

func (obj *CreateRoom) Identity() int {
	return CREATE_ROOM
}

func (obj *CreateRoom) Serialize() []byte {

	builder := flatbuffers.NewBuilder(0)
	builder.Finish(obj.create(builder))
	return builder.FinishedBytes()
}

func (obj *CreateRoom) Deserialize(bytes []byte) protocol.Protocol {
	root := source.GetRootAsCreateRoom(bytes, 0)
	return obj.parse(root)
}

type JoinRoom struct {
	Users []string
	Error uint32
}

func (obj *JoinRoom) users(builder *flatbuffers.Builder, users []string) flatbuffers.UOffsetT {
	_size := len(users)
	offsets := make([]flatbuffers.UOffsetT, _size)
	for i, x := range users {
		offsets[_size-i-1] = builder.CreateString(x)
	}

	builder.StartVector(4, _size, 4)
	for _, offset := range offsets {
		builder.PrependUOffsetT(offset)
	}
	return builder.EndVector(_size)
}

func (obj *JoinRoom) create(builder *flatbuffers.Builder) flatbuffers.UOffsetT {
	_users := obj.users(builder, obj.Users)

	source.JoinRoomStart(builder)
	source.JoinRoomAddUsers(builder, _users)
	source.JoinRoomAddError(builder, obj.Error)

	return source.JoinRoomEnd(builder)
}

func (obj *JoinRoom) parse(x *source.JoinRoom) *JoinRoom {

	_sizeUsers := x.UsersLength()
	obj.Users = make([]string, _sizeUsers)
	for i := 0; i < _sizeUsers; i++ {
		obj.Users = append(obj.Users, string(x.Users(i)))
	}
	obj.Error = x.Error()

	return obj
}

func (obj *JoinRoom) Identity() int {
	return JOIN_ROOM
}

func (obj *JoinRoom) Serialize() []byte {

	builder := flatbuffers.NewBuilder(0)
	builder.Finish(obj.create(builder))
	return builder.FinishedBytes()
}

func (obj *JoinRoom) Deserialize(bytes []byte) protocol.Protocol {
	root := source.GetRootAsJoinRoom(bytes, 0)
	return obj.parse(root)
}

type LeaveRoom struct {
	Id    string
	Error uint32
}

func (obj *LeaveRoom) create(builder *flatbuffers.Builder) flatbuffers.UOffsetT {
	_id := builder.CreateString(obj.Id)

	source.LeaveRoomStart(builder)
	source.LeaveRoomAddId(builder, _id)
	source.LeaveRoomAddError(builder, obj.Error)

	return source.LeaveRoomEnd(builder)
}

func (obj *LeaveRoom) parse(x *source.LeaveRoom) *LeaveRoom {
	obj.Id = string(x.Id())
	obj.Error = x.Error()

	return obj
}

func (obj *LeaveRoom) Identity() int {
	return LEAVE_ROOM
}

func (obj *LeaveRoom) Serialize() []byte {

	builder := flatbuffers.NewBuilder(0)
	builder.Finish(obj.create(builder))
	return builder.FinishedBytes()
}

func (obj *LeaveRoom) Deserialize(bytes []byte) protocol.Protocol {
	root := source.GetRootAsLeaveRoom(bytes, 0)
	return obj.parse(root)
}

type KickRoom struct {
	Success bool
	Error   uint32
}

func (obj *KickRoom) create(builder *flatbuffers.Builder) flatbuffers.UOffsetT {

	source.KickRoomStart(builder)
	source.KickRoomAddSuccess(builder, obj.Success)
	source.KickRoomAddError(builder, obj.Error)

	return source.KickRoomEnd(builder)
}

func (obj *KickRoom) parse(x *source.KickRoom) *KickRoom {
	obj.Success = x.Success()
	obj.Error = x.Error()

	return obj
}

func (obj *KickRoom) Identity() int {
	return KICK_ROOM
}

func (obj *KickRoom) Serialize() []byte {

	builder := flatbuffers.NewBuilder(0)
	builder.Finish(obj.create(builder))
	return builder.FinishedBytes()
}

func (obj *KickRoom) Deserialize(bytes []byte) protocol.Protocol {
	root := source.GetRootAsKickRoom(bytes, 0)
	return obj.parse(root)
}

type KickedRoom struct {
	Error uint32
}

func (obj *KickedRoom) create(builder *flatbuffers.Builder) flatbuffers.UOffsetT {

	source.KickedRoomStart(builder)
	source.KickedRoomAddError(builder, obj.Error)

	return source.KickedRoomEnd(builder)
}

func (obj *KickedRoom) parse(x *source.KickedRoom) *KickedRoom {
	obj.Error = x.Error()

	return obj
}

func (obj *KickedRoom) Identity() int {
	return KICKED_ROOM
}

func (obj *KickedRoom) Serialize() []byte {

	builder := flatbuffers.NewBuilder(0)
	builder.Finish(obj.create(builder))
	return builder.FinishedBytes()
}

func (obj *KickedRoom) Deserialize(bytes []byte) protocol.Protocol {
	root := source.GetRootAsKickedRoom(bytes, 0)
	return obj.parse(root)
}

type DestroyedRoom struct {
	Error uint32
}

func (obj *DestroyedRoom) create(builder *flatbuffers.Builder) flatbuffers.UOffsetT {

	source.DestroyedRoomStart(builder)
	source.DestroyedRoomAddError(builder, obj.Error)

	return source.DestroyedRoomEnd(builder)
}

func (obj *DestroyedRoom) parse(x *source.DestroyedRoom) *DestroyedRoom {
	obj.Error = x.Error()

	return obj
}

func (obj *DestroyedRoom) Identity() int {
	return DESTROYED_ROOM
}

func (obj *DestroyedRoom) Serialize() []byte {

	builder := flatbuffers.NewBuilder(0)
	builder.Finish(obj.create(builder))
	return builder.FinishedBytes()
}

func (obj *DestroyedRoom) Deserialize(bytes []byte) protocol.Protocol {
	root := source.GetRootAsDestroyedRoom(bytes, 0)
	return obj.parse(root)
}
